### PR TITLE
Add use thrall option

### PR DESCRIPTION
**Pull request by MarkKoester**
Adds use thrall toggle to "Extra Options" (#204)

Thrall DPT was taken from Bitterkoekje DPS spreadsheet (0.625 dps * 0.6 ticks per second = 0.375 dpt).

I wasn't sure if thrall dps should be included in any charts or ttk calculations or not so I left it out for now.

I'd be happy to add new tests if desired

Thanks for taking the time to look this over!

### CODE REVIEW

File: src/app/components/player/ExtraOptions.tsx
Comment (SuperNerdEric) [inline src/app/components/player/ExtraOptions.tsx:108]: These tooltips usually end in a period and I'm not sure if "in dps calculation" is necessary

**Comment by SuperNerdEric** [comment]
My gut tells me that the thrall DPS should be included in the charts since one of the use cases for actually including thralls in DPS calculation is to compare to a setup without thralls. Given that thralls bypass accuracy checks I could see it being useful for looking at the graph of Monster defence level vs DPS.


Also we might want to exclude places where thralls cannot be used:
- During the fight against [Galvek](https://oldschool.runescape.wiki/w/Galvek)
- During the fight against [Vanstrom Klause](https://oldschool.runescape.wiki/w/Vanstrom_Klause)
- During the fight against [Fragment of Seren](https://oldschool.runescape.wiki/w/Fragment_of_Seren)

There may be other places where thralls don't work for whatever reason. I guess Akkha enrage phase? But we don't have an easy way to check for that

**Comment by jayktaylor** [comment]
Thanks for the PR!

Agree with @SuperNerdEric here - it probably makes sense for us to include thralls' DPS in the actual graphs too.

Also, I think the tooltip is probably better off being "Thralls are summoned using the Arceuus spellbook.", which roughly matches what we tend to show in the tooltips of these extra options.

**Comment by MarkKoester** [comment]
Okay makes sense, I'll push the updated charts and wording soon and re-request an review. Thanks!

File: src/lib/CombatCalc.ts
Comment (MarkKoester) [inline src/lib/CombatCalc.ts:1612]: Previous condition doesn't match if you're splashing on an enemy so the ttk calculation never finishes

**Comment by MarkKoester** [comment]
I changed the ttk calculation to count ticks rather than hits. It then chooses a hit distribution to use based on who's attacking on that tick (no one, just thrall, just player, or both). Everything else stays pretty much the same.

I compared the thrall damage to a loadout with a max hit of 3 and 100% accuracy and the graphs are exactly the same, so it seems to be working as intended.

I'm still not sure what to do about the average ttk calculation. I feel like it should be similar to the ttk graph where it's based off of ticks rather than hits since hits don't really make sense when using thralls, but I'm not exactly sure how that should be implemented yet. In the meantime I just threw in an approximation by dividing health by dps, which is pretty close but not perfect

File: src/app/components/player/ExtraOptions.tsx
Comment (jayktaylor) [inline src/app/components/player/ExtraOptions.tsx:108]: Could you change this line specifically to
> Include thrall damage, from the Arceuus spellbook.
I think this avoids confusion for people that don't quite know what it is.

**Review by jayktaylor — COMMENTED** [review]
I'll let @LlemonDuck review the changes to the calc code before merging - thanks! 😄

File: src/app/components/player/ExtraOptions.tsx
Comment (LlemonDuck) [inline src/app/components/player/ExtraOptions.tsx:111]: Should this just be Toggle's help attr?

File: src/lib/CombatCalc.ts
Comment (LlemonDuck) [inline src/lib/CombatCalc.ts:48]: This is far less than the original for low-speed weapons. Let's use `TTK_DIST_MAX_ITER_ROUNDS * weaponSpeed` as the upper bound instead and keep this at 1000.

File: src/lib/CombatCalc.ts
Comment (LlemonDuck) [inline src/lib/CombatCalc.ts:62]:
```ts
const THRALL_DIST = HitDistribution.linear(1.0, THRALL_MAX);
```

File: src/lib/CombatCalc.ts
Comment (LlemonDuck) [inline src/lib/CombatCalc.ts:67]: Either we should not use zip here or we should cache the result of zip whenever possible. Definitely don't recompute it on every tick that both attacks happen.

zip is one of the worst-case time complexity things we can do in the ttk dist.

It's probably actually better to not zip at all, and instead just apply both dists separately since they're not correlated. ```

File: src/lib/CombatCalc.ts
Comment (LlemonDuck) [inline src/lib/CombatCalc.ts:1682]: Leave a comment here stating that this doesn't account for thrall projectile travel time.

File: src/lib/CombatCalc.ts
Comment (LlemonDuck) [inline src/lib/CombatCalc.ts:1676]: If you're going to -1 everything everywhere you may as well keep this starting at 0. Just add one for the map get/sets.

File: src/lib/CombatCalc.ts
Comment (LlemonDuck) [inline src/lib/CombatCalc.ts:61]: I like to preserve the math in constants so it's more clear how they were derived:
```ts
const THRALL_MAX = 3;
const THRALL_SPEED = 4;
const THRALL_DPT = sum(range(0, THRALL_MAX + 1)) / ((THRALL_MAX + 1) * THRALL_SPEED);
```

File: src/lib/CombatCalc.ts
Comment (LlemonDuck) [inline src/lib/CombatCalc.ts:1635]: getHtk() is never used anymore, remove it

**Review by LlemonDuck — CHANGES_REQUESTED** [review]

File: src/app/components/player/ExtraOptions.tsx
Comment (MarkKoester) [reply src/app/components/player/ExtraOptions.tsx:111]: It seems like keeping it in the label is better if you want to keep the toggle component general purpose. You may want to remove the help completely since it's not used anywhere

**Comment by LlemonDuck** [comment]
closing in favor of #229

**Pull request closed without merging**
